### DISCUSSION TITLE
Contributing IdentityProvider for German GesundheitsID (eHealthID)

### DISCUSSION THREAD

**Discussion by thomasrichner-oviva** [keycloak/keycloak#26447] [Ideas]
Hey all,

I'd like to contribute an IdentityProvider for the German GesundheitsID, which is Germany's eHealth Identity.

After diving through the IdentityProvider interface and existing implementations of other providers I struggle a bit on how to best implement the broker flow on the Keycloak side and it would be great if I can get some pointers:

There are two cases:
1. A Keycloak account exists with no credentials yet. A trusted client can hand out a token to the user to link a given account. I.e. they signed up, with have an ephemeral session in some other service and would like the user to link the external identity without ever having to set up local credentials.
2. A Keycloak account exists with credentials. At some later point the user would like to link the external identity. There is no attribute like email to link an account, so the way to link it is to proof that one has credentials for both accounts.

With that in mind, I struggle with the following questions:
1. How do I control linking of accounts where a existing user can verify the identity link by going through an existing authentication flow (i.e. login with password & 2FA).
     1.  How to prevent non-existing accounts from being created?
     2. How to prevent magic account linking based on email etc. ?
3. How do I enable a trusted client to initiate the linking of an existing account, some action token to start an idp link flow with an admin API to issue such tokens?

**Comment by jonkoops** [comment]
This sounds like something that could be part of an extension, rather than building this into Keycloak itself. If there are things missing in our public APIs that would help implement this, feel free to open issues and PRs around improving that.

**Reply by thomasrichner-oviva** [reply]
Thank's for the pointers, happy to do that as an extension.

Just to verify my assumptions:

#### Case 1
> A Keycloak account exists with credentials. At some later point the user would like to link the external identity. There is no attribute like email to link an account, so the way to link it is to proof that one has credentials for both accounts.

I assume a custom IdentityProvider + a first-login flow can do that, right?

#### Case 2
> A Keycloak account exists with no credentials yet. A trusted client can hand out a token to the user to link a given account. I.e. they signed up, with have an ephemeral session in some other service and would like the user to link the external identity without ever having to set up local credentials.

How would I do that with an extension? Is there a similar example?

**Reply by jonkoops** [reply]
> I assume a custom IdentityProvider + a first-login flow can do that, right?

Yes, I think that would cover it.

> How would I do that with an extension? Is there a similar example?

I think you could accomplish this with the existing Authentication SPIs, but I am not too familiar with this. I'd recommend doing some research on which SPIs are out there to accomplish this.

**Reply by thomasdarimont** [reply]
Case 1) Would work as you describe.

Case 2) If you know the userIds / usernames in both systems, you could use a privileged service client to set the linked identity directly via Realm Rest API via: `org.keycloak.services.resources.admin.UserResource#addFederatedIdentity(..)`. Once the IdP link is established, you only need to login the user via the now connected IdP, e.g. through a dynamic redirect.

**Reply by thomasrichner-oviva** [reply]
Ah, thanks for the hint regarding case 2) . That might be an option indeed.

**Comment by thomasrichner-oviva** [comment]
To close the loop, we've implemented a relying-party instead that can be easily used with Keycloak: https://github.com/oviva-ag/ehealthid-relying-party
